rotkehlchen/assets/utils.py

```diff
@@ -33,7 +34,81 @@ def add_ethereum_token_to_db(token_data: EvmToken) -> EvmToken:
         data=token_data,
     )
     # This can, but should not raise UnknownAsset, DeserializationError
-    return EvmToken(token_data.identifier, form_with_incomplete_data=True)
+    return EvmToken(token_data.identifier)
+
+
+def _query_token_information(
+        evm_inquirer: 'EvmNodeInquirer',
+        evm_address: ChecksumEvmAddress,
+        name: Optional[str],
+        symbol: Optional[str],
+        decimals: Optional[int],
+        token_kind: EvmTokenKind,
+) -> tuple[Optional[str], Optional[str], Optional[int]]:
+    """
+    Query ethereum to retrieve basic contract information for the given address
+    May raise:
+    - NotERC20Conformant
+    """
+    if token_kind == EvmTokenKind.ERC20:
+        info = evm_inquirer.get_erc20_contract_info(evm_address)
+        decimals = info['decimals'] if decimals is None else decimals
+        symbol = info['symbol'] if symbol is None else symbol
+        name = info['name'] if name is None else name
+        if None in (decimals, symbol, name):
+            raise NotERC20Conformant(f'Token {evm_address} is not ERC20 conformant')  # noqa: E501  # pylint: disable=raise-missing-from
+
+    elif token_kind == EvmTokenKind.ERC721:
+        info = evm_inquirer.get_erc721_contract_info(evm_address)
+        decimals = 0
+        if symbol is None:
+            symbol = info['symbol'] if info['symbol'] is not None else ''
+        if name is None:
+            name = info['name'] if info['name'] is not None else ''
+
+    else:
+        raise NotERC20Conformant(f'Token {evm_address} is of uknown type')  # pylint: disable=raise-missing-from  # noqa: E501
+
+    return name, symbol, decimals
+
+
+def _edit_token_and_clean_cache(
+        ethereum_token: EvmToken,
+        name: Optional[str],
+        decimals: Optional[int],
+        evm_inquirer: Optional['EvmNodeInquirer'],
+) -> None:
+    """
+    Update information regarding name and decimals for an ethereum token.
+    If name is missing in the database and is not provided then query the blockchain for it
+    """
+    updated_fields = False
+
+    if ethereum_token.name == ethereum_token.identifier and name is not None:
+        object.__setattr__(ethereum_token, 'name', name)
+        updated_fields = True
+    elif ethereum_token.name == ethereum_token.identifier and name is None and evm_inquirer is not None:  # noqa: E501
```

Comment:
you are repeating the if check here and checking two things again
```suggestion
    if ethereum_token.name == ethereum_token.identifier:
        if name is not None:
            object.__setattr__(ethereum_token, 'name', name)
            updated_fields = True
        elif evm_inquirer is not None:  # noqa: E501
            ...
```